File: backend/rss.py
import sqlite3
from time import gmtime, strptime
from calendar import timegm
import feedparser

feed_url = "https://www.gdacs.org/xml/rss.xml"
db_file = "feed.db" 
database = sqlite3.connect(db_file, check_same_thread=False) # python's 3.11's sqlite3 is compiled with threadsafety already

def create_entries_table():
    #TODO: NEEDS reasonable schema here

    # initialize table
    entries_table = '''
                CREATE TABLE IF NOT EXISTS disasters (
                    disasterID TEXT PRIMARY KEY,
                    name TEXT,
                    type TEXT NOT NULL,
                    eventID TEXT NOT NULL,
                    fromdate INTEGER NOT NULL,
                    todate INTEGER NOT NULL
                );

                CREATE TABLE IF NOT EXISTS entries (
                    timestamp INTEGER NOT NULL,
                    title TEXT NOT NULL,
                    disasterID TEXT NOT NULL,
                    alertLevel TEXT NOT NULL,
                    summary TEXT,
                    countries TEXT,
                    latitude REAL NOT NULL,
                    longitude REAL NOT NULL,
                    PRIMARY KEY (timestamp, title),
                    FOREIGN KEY (disasterID) REFERENCES disasters(disasterID)
                );


                '''
    try:
        cursor = database.cursor()
        cursor.executescript(entries_table)
        database.commit()
    
    except sqlite3.Error as e:
        print(e)

def populate_entries():
    try:
        newsfeed = feedparser.parse(feed_url)
        cursor = database.cursor()
        insert_entry_statement = '''
                                INSERT OR IGNORE
                                INTO entries(timestamp, title, disasterID, alertLevel, summary, countries, latitude, longitude)
                                VALUES(?,?,?,?,?,?,?,?);
                                '''

        upsert_disaster_statement = '''
                                    INSERT INTO disasters(disasterID, name, type, eventID, fromdate, todate) 
                                    VALUES(?,?,?,?,?,?)
                                    ON CONFLICT(disasterID) DO UPDATE SET
                                        todate = excluded.todate
                                    WHERE excluded.todate > disasters.todate
                                    '''
        for entry in newsfeed.entries:
            # entries table's value
            disasterid = entry.id
            unixTimestamp = int(timegm(entry.published_parsed))
            title = entry.title
            alertLevel = entry.gdacs_alertlevel
            summary = entry.summary
            countries = entry.gdacs_country # comma separated if multiple
            lat = entry.geo_lat
            long = entry.geo_long

            # disasters table's value
            name = entry.gdacs_eventname
            typ = entry.gdacs_eventtype
            eventID = entry.gdacs_eventid 
            fromdate = int(timegm(strptime(entry.gdacs_fromdate, '%a, %d %b %Y %H:%M:%S GMT')))
            todate = int(timegm(strptime(entry.gdacs_todate, '%a, %d %b %Y %H:%M:%S GMT')))

            cursor.execute(insert_entry_statement, (unixTimestamp, 
                                                    title,
                                                    disasterid,
                                                    alertLevel,
                                                    summary, 
                                                    countries,
                                                    lat,
                                                    long
                                                    ))
            cursor.execute(upsert_disaster_statement, (disasterid,
                                                       name,
                                                       typ,
                                                       eventID,
                                                       fromdate,
                                                       todate
                                                       ))

            database.commit()

    except sqlite3.Error as e:
        print(e, " from popEn")

def generate_proximity(lat, long, prox):
    '''calculate raw proximity, then normalize the result'''

    # upper and lower limit to make sure range stays sane as coordinates wrap all the way around
    lat_upper_limit = lat
    lat_lower_limit = lat 
    long_upper_limit = long
    long_lower_limit = long

    lat_start = lat - prox
    lat_end = lat + prox

    # normalize latitude
    if lat_start < -90:
        lat_start += 180
        lat_upper_limit = 90
        lat_lower_limit = -90
    elif lat_end >= 90:
        lat_end -= 180
        lat_upper_limit = 90
        lat_lower_limit = -90
    
    long_start = long - prox
    long_end = long + prox
    
    # normalize longitude
    if long_start < -180:
        long_start += 360
        long_upper_limit = 180
        long_lower_limit = -180
    elif long_end >= 180:
        long_end -= 360
        long_upper_limit = 180
        long_lower_limit = -180

    return (lat_start, lat_upper_limit, lat_lower_limit, lat_end, long_start, long_upper_limit, long_lower_limit, long_end)

def query(latitude, longitude):
    cursor = database.cursor()

    lat = float(latitude)
    long = float(longitude)
    
    # TC range is +-3, otherwise +-1
    sql_inputs = generate_proximity(lat, long, 3) + generate_proximity(lat, long, 1) 

    query = '''
                SELECT * FROM (entries LEFT JOIN disasters ON entries.disasterID = disasters.disasterID) 
                WHERE CASE  
                WHEN type = "TC" THEN
                ((latitude BETWEEN ? AND ?) OR (latitude BETWEEN ? AND ?)) AND ((longitude BETWEEN ? AND ?) OR (longitude BETWEEN ? AND ?)) 
                ELSE
                ((latitude BETWEEN ? AND ?) OR (latitude BETWEEN ? AND ?)) AND ((longitude BETWEEN ? AND ?) OR (longitude BETWEEN ? AND ?))
                END
                ORDER BY timestamp DESC;
            '''
    cursor.execute(query, sql_inputs)

    headers = list(map(lambda attr : attr[0], cursor.description))
    results = [{header:row[i] for i, header in enumerate(headers)} for row in cursor]

    # converting all timestamps to ISO format
    # split countries into a list for easy access
    return results


def bootstrapping():
    create_entries_table()
    populate_entries()

File: backend/internal/rss/rss.go
package rss

import (
	"aggregate-site/backend/internal/database"
	"strconv"
	"strings"
	"time"

	"github.com/mmcdole/gofeed"
	ext "github.com/mmcdole/gofeed/extensions"
)

var feedURL string = "https://www.gdacs.org/xml/rss.xml"

func findKeyValue(ext map[string][]ext.Extension, keys []string) string {

	// pop the first key
	key, remainingKeys := keys[0], keys[1:]

	valArr := ext[key]
	// for now there is no multiple extensions in one key,
	// so we just take first item
	val := valArr[0]
	if len(remainingKeys) == 0 {
		return val.Value
	}
	return findKeyValue(val.Children, remainingKeys)
}

func findExtensionValue(item *gofeed.Item, key string) string {
	keys := strings.Split(key, ":")
	namespace, keys := keys[0], keys[1:]

	ext := item.Extensions[namespace] // find namespace
	return findKeyValue(ext, keys)
}

func Fetch(db database.Database) error {
	var err error

	fp := gofeed.NewParser()
	feed, _ := fp.ParseURL(feedURL)

	items := feed.Items

	entries := []database.Entry{}
	disasters := []database.Disaster{}
	// TODO:rewrite this for concurrency
	for _, item := range items {
		lat, err := strconv.ParseFloat(findExtensionValue(item, "geo:Point:lat"), 64)
		if err != nil {
			return err
		}
		long, err := strconv.ParseFloat(findExtensionValue(item, "geo:Point:long"), 64)
		if err != nil {
			return err
		}

		// format the entry
		var entry = database.Entry{
			DisasterID: item.GUID,
			Title:      item.Title,
			Summary:    item.Description,
			Timestamp:  item.PublishedParsed.Unix(),
			AlertLevel: findExtensionValue(item, "gdacs:episodealertlevel"),
			Countries:  findExtensionValue(item, "gdacs:country"),
			Latitude:   lat,
			Longitude:  long,
		}
		entries = append(entries, entry)

		fromDate, err := time.Parse(
			time.RFC1123,
			findExtensionValue(item, "gdacs:fromdate"),
		)
		if err != nil {
			return err
		}

		toDate, err := time.Parse(
			time.RFC1123,
			findExtensionValue(item, "gdacs:todate"),
		)
		if err != nil {
			return err
		}

		// format the disaster
		var disaster = database.Disaster{
			DisasterID: item.GUID,
			Name:       findExtensionValue(item, "gdacs:eventname"),
			Type:       findExtensionValue(item, "gdacs:eventtype"),
			EventID:    findExtensionValue(item, "gdacs:eventid"),
			FromDate:   fromDate.Unix(),
			ToDate:     toDate.Unix(),
		}
		disasters = append(disasters, disaster)
	}
	err = db.InsertEntries(entries)
	if err != nil {
		return err
	}
	err = db.InsertDisaster(disasters)
	return err
}
